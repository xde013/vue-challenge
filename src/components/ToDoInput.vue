<template>
  <div class="todo-input">
      <input class="text-field" @keyup.enter='addTodo()' placeholder="Take a note" v-model="text" type="text"/>
      <button class="submit-button" :disabled='!addTodoPossible' @click="addTodo()">+</button>
  </div>
</template>

<script>
export default {
  name: 'ToDoInput',
  data () {
    return {
      text: ''
    }
  },
  computed: {
    addTodoPossible () {
      return this.text && this.text !== ''
    }
  },
  methods: {
    addTodo () {
      this.$store.dispatch('addTodo', {text: this.text, checked: false, createdAt: new Date()})
      this.text = ''
    }
  }}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="stylus">
    .todo-input
        display flex
        border-radius 2px
        padding 2px 8px
        background-color: rgb(250, 250, 250);
        box-shadow: 0 3px 5px rgba(0,0,0,0.20);
        font-size: 14px;
        line-height: 19px;
        padding: 12px;
        .text-field
            border none
            outline-width: 0;
            background-color transparent
            flex 1
        .submit-button
            color: rgba(0,0,0,.7)
            font-weight 300
            outline-width: 0;
            border none
            background-color transparent
            cursor pointer
            border-radius 2px
            font-size .8rem
</style>
