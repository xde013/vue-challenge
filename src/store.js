import Vue from 'vue'
import Vuex from 'vuex'

Vue.use(Vuex)

export const mutations = {
  addTodo: (state, payload) => {
    state.todos.unshift(payload)
  },
  updateTodo: (state, payload) => {
    // TODO
  },
  removeTodo: (state, payload) => {
    // TODO
  }
}

export const actions = {
  addTodo: ({commit}, payload) => {
    commit('addTodo', payload)
  },
  updateTodo: ({commit}, payload) => {
    commit('updateTodo', payload)
  },
  removeTodo: ({commit}, payload) => {
    commit('removeTodo', payload)
  }
}

export const defaultState = {
  todos: [
    {
      text: 'Think real hard about whats for lunch',
      checked: false,
      createdAt: new Date(Date.now() - 3600)
    },
    {
      text: 'Buy some Pantos and start looking for Lambos',
      checked: false,
      createdAt: new Date(Date.now() - 7200)
    }
  ]
}

export default new Vuex.Store({
  state: defaultState,
  mutations,
  actions
})
