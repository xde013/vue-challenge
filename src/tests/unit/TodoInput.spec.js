import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import ToDoInput from '@/components/ToDoInput.vue'
import { defaultState, getters, mutations, actions } from '@/store'

const localVue = createLocalVue()
localVue.use(Vuex)

const factory = (options = {}) => {
  return shallowMount(ToDoInput, options)
}

describe('ToDoList.vue', () => {
  let store
  let localConfig
  beforeEach(() => {
    store = new Vuex.Store({
      state: defaultState,
      getters,
      mutations,
      actions
    })
    localConfig = { store, localVue }
  })

  it('dispatch addToDo when button is clicked', () => {
    const mockStore = { dispatch: jest.fn() }
    const wrapper = factory({mocks: { $store: mockStore }})
    const textInput = wrapper.find('input[type="text"]')
    textInput.setValue('Buy Milk')
    wrapper.find('.submit-button').trigger('click')
    expect(mockStore.dispatch).toHaveBeenCalled()
  })
  it('do not allow disptach addTodo when text is empty', () => {
    const mockStore = { dispatch: jest.fn() }
    const wrapper = factory({mocks: { $store: mockStore }})
    const textInput = wrapper.find('input[type="text"]')
    textInput.setValue('')
    wrapper.find('.submit-button').trigger('click')
    expect(mockStore.dispatch).not.toHaveBeenCalled()
  })
  it('matches snapshot', () => {
    const wrapper = shallowMount(ToDoInput, localConfig)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
