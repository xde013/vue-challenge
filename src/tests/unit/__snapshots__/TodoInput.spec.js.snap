// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList.vue matches snapshot 1`] = `
<div class="todo-input">
  <input placeholder="Take a note" type="text" class="text-field">
  <button disabled="disabled" class="submit-button">+</button>
</div>
`;
